src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,89 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
+    private static final Map<Class<?>, Class<?>> IMPLEMENTATIONS = new HashMap<>();
+
+    static {
+        IMPLEMENTATIONS.put(ProductService.class, ProductServiceImpl.class);
+        IMPLEMENTATIONS.put(ProductParser.class, ProductParserImpl.class);
+        IMPLEMENTATIONS.put(FileReaderService.class, FileReaderServiceImpl.class);
+    }
+
+    private static final Map<Class<?>, Object> INSTANCES = new HashMap<>();
     private static final Injector injector = new Injector();
 
     public static Injector getInjector() {
         return injector;
     }
 
-    public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+    public Object getInstance(Class<?> interfaceClass) {
+        Object classImplementationInstance = null;
+        Class<?> implementationClass = findImplementation(interfaceClass);
+        Field[] classFields = implementationClass.getDeclaredFields();
+        for (Field field : classFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                classImplementationInstance = createNewInstance(implementationClass);
+
+                try {
+                    field.setAccessible(true);
+                    field.set(classImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException(
+                            "Cannot set field '" + field.getName()
+                                    + "' in class '" + implementationClass.getName() + "'.", e);
+                }
+            }
+        }
+        if (classImplementationInstance == null) {
+            classImplementationInstance = createNewInstance(implementationClass);
+        }
+        return classImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> implementationClass) {
+        if (INSTANCES.containsKey(implementationClass)) {
+            return INSTANCES.get(implementationClass);
+        }
+        try {
+            Constructor<?> constructor = implementationClass.getConstructor();
+            Object instance = constructor.newInstance();
+            INSTANCES.put(implementationClass, instance);
+            return instance;
+        } catch (NoSuchMethodException | InvocationTargetException
+                | InstantiationException | IllegalAccessException e) {
```

Comment:
can we replace this 4 exceptions with common parent?

Reply:
Right, have to read out task conditions more precisely )